src/transformers/configuration_utils.py

```diff
@@ -422,6 +427,24 @@ def num_labels(self, num_labels: int):
             self.id2label = {i: f"LABEL_{i}" for i in range(num_labels)}
             self.label2id = dict(zip(self.id2label.values(), self.id2label.keys()))
 
+    @property
+    def attn_implementation(self):
+        if hasattr(self, "_attn_implementation"):
```

Comment:
It can not be an attribute in case `super().__init__` is not called from a child class `__init__`, a test was failing due to this:
https://github.com/huggingface/transformers/blob/2098d343cc4b4b9d2aea84b3cf1eb5a1e610deff/tests/test_modeling_utils.py#L356
https://github.com/huggingface/transformers/blob/2098d343cc4b4b9d2aea84b3cf1eb5a1e610deff/utils/test_module/custom_configuration.py#L12
https://github.com/huggingface/transformers/blob/2098d343cc4b4b9d2aea84b3cf1eb5a1e610deff/utils/test_module/custom_modeling.py#L22 (I don't understand why `NoSuperInitModel` calls `super().__init__` though)

Reply:
Can we move this property to `PreTrainedModel`?

Reply:
I don't think it is possible. Essentially, moving this property would require to:

1. Keep the property `_flash_attn_2_enabled` defined in `from_pretrained` as this one is accessible nowhere else than in `config` in the `XXXDecoderLayer` init: https://github.com/huggingface/transformers/blob/fdd86eed3b5bad8cf3da339fb3acb53fa13eb74b/src/transformers/models/llama/modeling_llama.py#L634
2. Move the checks/dispatch of eager/SDPA/flash in `XXXDecoderLayer` init instead of having in `from_pretrained`/`from_config`/PreTrainedModel `__init__`, but this is quite difficult as we have the inheritance `LlamaDecoderLayer(nn.Module)` and so attributes as `_supports_flash_attn_2`, `use_bettertransformer`, `_supports_sdpa` are not accessible from `self`. So it's impossible to check without changing the inheritance (or moving `_supports_flash_attn_2` at the file level - which is not good IMO)

Anyway, we would remove `config.attn_implementation`, but this implies adding back `config._flash_attn_2_enabled`. I don't think it is better.

What I can propose @ArthurZucker @patrickvonplaten is to make `attn_implementation` fully private in `PretrainedConfig`, and never obeyed to. Does that sound fine?

Reply:
Made it private!